### PR TITLE
Xamarin.Mac backend (dev branch)

### PR DESCRIPTION
**Pull request by mhutch**
Adds a Xamarin.Mac backend and test app for it.

### CODE REVIEW

File: src/ADAL.PCL.Mac/AuthenticationAgentNSWindowController.cs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        {
+            if (request == null)
+            {
+                WebView.DecideUse(decisionToken);
+                return;
+            }
+
+            string requestUrlString = request.Url.ToString();
+
+            if (requestUrlString.StartsWith(BrokerConstants.BrowserExtPrefix, StringComparison.OrdinalIgnoreCase))
```
Comment (kpanwar) [inline src/ADAL.PCL.Mac/AuthenticationAgentNSWindowController.cs]: launching browser in Mac for device enrollment is not  a supported scenario. native objc library returns error with "server is redirecting us to browser, this behavior is not defined on Mac OS X yet" message

File: src/ADAL.PCL.Mac/PlatformParameters.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// Skips calling to broker if broker is present. false, by default
+        /// </summary>
+        public bool UseBroker { get; set; }
+
+        /// <summary>
+        /// Shows the prompt in a modal dialog.
+        /// </summary>
+        public bool UseModalDialog { get; set; }
```
Comment (kpanwar) [inline src/ADAL.PCL.Mac/PlatformParameters.cs]: native library does not provide an option to use modal dialog. AFAIK, it forces modal by default.

File: src/ADAL.PCL.Mac/PlatformParameters.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// Caller NSWindow
+        /// </summary>
+        public NSWindow CallerWindow { get; private set; }
+
+        /// <summary>
+        /// Skips calling to broker if broker is present. false, by default
+        /// </summary>
+        public bool UseBroker { get; set; }
```
Comment (kpanwar) [inline src/ADAL.PCL.Mac/PlatformParameters.cs]: Broker is not supported at all. We should not expose this flag.

File: src/ADAL.PCL.Mac/PlatformParameters.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+//------------------------------------------------------------------------------
+
+using AppKit;
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory
+{
+    /// <summary>
+    /// Additional parameters used in acquiring user's authorization
+    /// </summary>
+    public class PlatformParameters : IPlatformParameters
```
Comment (kpanwar) [inline src/ADAL.PCL.Mac/PlatformParameters.cs]: Just as an FYI, we would also need to support PromptBehavior enumeration in Mac. I don't expect you to do this work. I will add it once your PR is merged and will cut a release to include the same.

File: src/ADAL.PCL.iOS/BrokerKeyHelper.cs [hunk 5]
```diff
@@ -73,7 +73,12 @@ internal static byte[] GetRawBrokerKey()
                     ValueData = byteData
                 };
 
-                SecStatusCode code = SecKeyChain.Add(record);
+                var result = SecKeyChain.Add(record);
```
Comment (kpanwar) [inline src/ADAL.PCL.iOS/BrokerKeyHelper.cs]: we don't need brokerkeyhelper class here at all.

File: src/ADAL.PCL.iOS/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -39,7 +39,4 @@
 // The following GUID is for the ID of the typelib if this project is exposed to COM
 [assembly: Guid("ff47962a-d498-4c63-b7e9-4db3653ad7e0")]
 
-// Assembly version information is in file ADAL.Common\CommonAssemblyInfo.cs
-
-[assembly: AssemblyDelaySign(true)]
-[assembly: AssemblyKeyFile(@"..\..\build\35MSSharedLib1024.snk")]
```
Comment (kpanwar) [inline src/ADAL.PCL.iOS/Properties/AssemblyInfo.cs]: we require delay signing for our libraries. any specific reason this was removed?

File: src/ADAL.PCL.iOS/TokenCachePlugin.cs [hunk 7]
```diff
@@ -33,6 +33,12 @@ namespace Microsoft.IdentityModel.Clients.ActiveDirectory
 {
     internal class TokenCachePlugin : ITokenCachePlugin
     {
+#if MAC
```
Comment (kpanwar) [inline src/ADAL.PCL.iOS/TokenCachePlugin.cs]: in native library we do not persist the cache in keychain when using Mac. the persistence is completely owned by the developer by implemeting the delegates

File: src/ADAL.PCL/WsTrustRequest.cs [hunk 8]
```diff
@@ -78,16 +78,6 @@ internal static class WsTrustRequest
             }
 
             StringBuilder messageBuilder = BuildMessage(DefaultAppliesTo, wsTrustAddress, credential);
-            string soapAction = XmlNamespace.Issue.ToString();
```
Comment (kpanwar) [inline src/ADAL.PCL/WsTrustRequest.cs]: why was this code removed? We need this for integrated auth support.

File: src/ADAL.PCL.iOS/BrokerKeyHelper.cs (hunk 5 above)
Comment (mhutch) [inline src/ADAL.PCL.iOS/BrokerKeyHelper.cs]: This isn't used in the Mac backend. It was causing a warn-as-error build failure in the iOS backend.

File: src/ADAL.PCL.iOS/Properties/AssemblyInfo.cs (hunk 6 above)
Comment (mhutch) [inline src/ADAL.PCL.iOS/Properties/AssemblyInfo.cs]: This was failing when building on Mac. I moved it to the csproj instead.

File: src/ADAL.PCL/WsTrustRequest.cs (hunk 8 above)
Comment (mhutch) [inline src/ADAL.PCL/WsTrustRequest.cs]: It was dead code, causing warning-as-error build failures.

File: src/ADAL.PCL.Mac/PlatformParameters.cs (hunk 2 above)
Comment (mhutch) [inline src/ADAL.PCL.Mac/PlatformParameters.cs]: From what I can see reading the source of the obj-c version, the dialog isn't actually modal, more like a singleton. This seems a little odd.

File: src/ADAL.PCL.Mac/PlatformParameters.cs (hunk 3 above)
Comment (mhutch) [inline src/ADAL.PCL.Mac/PlatformParameters.cs]: Oh, my mistake. I assumed it was an interface member.

File: src/ADAL.PCL.iOS/TokenCachePlugin.cs (hunk 7 above)
Comment (mhutch) [inline src/ADAL.PCL.iOS/TokenCachePlugin.cs]: Is there a reason for this or was it simply never implemented?

File: src/ADAL.PCL.Mac/AuthenticationAgentNSWindowController.cs (hunk 1 above)
Comment (mhutch) [inline src/ADAL.PCL.Mac/AuthenticationAgentNSWindowController.cs]: Fixed!

**Comment by kpanwar** [comment]
:shipit:

**Pull request merged**